# Concatenar y Apendizar data sets

In [ ]:
import pandas as pd
import numpy as np
from os import path, mkdir, getcwd

In [ ]:
red_wine = pd.read_csv("https://raw.githubusercontent.com/joanby/python-ml-course/master/datasets/wine/winequality-red.csv", sep = ";")
white_wine = pd.read_csv("https://raw.githubusercontent.com/joanby/python-ml-course/master/datasets/wine/winequality-white.csv", sep = ";")

In [ ]:
red_wine.head()

In [ ]:
white_wine.head()

In [ ]:
red_wine.columns.values

In [ ]:
white_wine.columns.values

In [ ]:
print(f"en el dataset hay {red_wine.shape} vino tinto y {white_wine.shape} vino blanco")

## Concatenacion
para concatenar utilizamos la funcion: `pd.concat` <br>
`pd.concat([dataframe1, dataframe2], axis=0)`

Recordatorio:
En python tenemos dos tipos de ejes:

    1. axis=0 denota el eje vertical
    2. axis=1 denota el eje horizontal

Lo que hacemos con esto es apiilar uno arriba del otro, el que ponemos primero en la lista es el que ira primero.

In [ ]:
wine_data = pd.concat([red_wine, white_wine], axis = 0)

In [ ]:
wine_data.shape

In [ ]:
wine_data.head()

In [ ]:
wine_data.tail()

In [ ]:
data1 = wine_data.head(10)
data2 = wine_data[300:310]
data3 = wine_data.tail(10)

In [ ]:
wine_scranble = pd.concat([data1, data2, data3], axis = 0)

In [ ]:
wine_scranble

In [ ]:
wine_scranble = pd.concat([data3, data2, data1], axis = 0, )
wine_scranble

si agregamos el parametro `ignore_index=True`, como lo dice ignorara y ordenara los id de cada fila sin importar antes como aparecian.

## Datos Distribuidos.

Vamos a trabajar con una gran cantidad de datos, lo primero es descargarlos, y para fijarnos las estructura cargamos el primero.

Podiamos hacer todo el segundo paso desde un inicio pero si desea tener cada dataset siga el siguiente paso.

### Descarga de dataset.

In [ ]:
url = 'https://raw.githubusercontent.com/joanby/python-ml-course/master/datasets/distributed-data/'
data = pd.read_csv(url+"001.csv", sep=',')
data.head()



In [ ]:
for i in range(1, 333):

    if i<10:
        i = '00' + str(i)
    elif i<100:
        i = '0' + str(i)
    else:
        i = str(i)
    
    url_data = url + i + '.csv'
    print(url_data)
    data = pd.read_csv(url_data, sep = ",")
    data.to_csv('../data/distributed_data/' + i + '.csv')

### Concatenamos los datos

In [ ]:
filepath = "../data/distributed_data/"

data = pd.read_csv("../data/distributed_data/001.csv", sep = ",")
final_length = len(data)

In [ ]:

for i in range(2, 333):
    if i<10:
        filename = '00' + str(i)
    elif i<100:
        filename = '0' + str(i)
    else:
        filename = str(i)
        
    file = filepath + filename + '.csv'
    temp_data = pd.read_csv(file, sep = ",")
    data = pd.concat([data, temp_data], axis = 0)

    final_length += len(temp_data)
#data.to_csv('../data/distributed_data/distributed data complete.csv')

In [ ]:
data.shape

Comprobamos que el numero de filas final es igual al numero de filas que agregamos, con `final_leangth=len(data)` contamos cada fila, y `data.shape[0]` nos data el numero total de filas.

In [ ]:
final_length == data.shape[0]

## Join de datasets

In [18]:
filpath = 'https://raw.githubusercontent.com/joanby/python-ml-course/master/datasets/athletes/Medals.csv'
data_main = pd.read_csv(filpath, encoding='ISO-8859-1')
data_main.to_csv("../data/olympic/medals.csv")

In [19]:
data_main.head()

Athlete   Age  Year Closing Ceremony Date  Gold Medals  \
0    Michael Phelps  23.0  2008            08/24/2008            8   
1    Michael Phelps  19.0  2004            08/29/2004            6   
2    Michael Phelps  27.0  2012            08/12/2012            4   
3  Natalie Coughlin  25.0  2008            08/24/2008            1   
4     Aleksey Nemov  24.0  2000            10/01/2000            2   

   Silver Medals  Bronze Medals  Total Medals  
0              0              0             8  
1              0              2             8  
2              2              0             6  
3              2              3             6  
4              1              3             6

In [52]:
a = data_main["Athlete"].unique().tolist()
len(a)

6956

Atletas unicos seleccionados hay 6956

In [21]:
data_main.shape

(8618, 8)

En el dataset original hay 8618, es decir hay 1662 duplicados.

In [22]:
data_country = pd.read_csv('https://raw.githubusercontent.com/joanby/python-ml-course/master/datasets/athletes/Athelete_Country_Map.csv', encoding='ISO-8859-1')
data_country.head()

Athlete        Country
0    Michael Phelps  United States
1  Natalie Coughlin  United States
2     Aleksey Nemov         Russia
3     Alicia Coutts      Australia
4    Missy Franklin  United States

In [23]:
len(data_country)

6970

In [28]:
data_sports = pd.read_csv('https://raw.githubusercontent.com/joanby/python-ml-course/master/datasets/athletes/Athelete_Sports_Map.csv', encoding='ISO-8859-1')
data_sports.head()
data_sports.to_csv("../data/olympic/Athelete_Sports_Map.csv")


In [29]:
len(data_sports)

6975

In [35]:
data_sports[
    (data_sports["Athlete"] == "Matt Ryan") | 
    (data_sports["Athlete"] == "Richard Thompson") | 
    (data_sports["Athlete"] == "Chen Jing")
]

Athlete         Sport
528   Richard Thompson     Athletics
1308         Chen Jing    Volleyball
1419         Chen Jing  Table Tennis
2727         Matt Ryan        Rowing
5003         Matt Ryan    Equestrian
5691  Richard Thompson      Baseball

In [37]:
'''
data_main_country = pd.merge(left = data_main, right = data_country,
                            left_on = "Athlete", right_on = "Athlete")
'''

In [38]:
data_main_country.head()

Athlete   Age  Year Closing Ceremony Date  Gold Medals  \
0    Michael Phelps  23.0  2008            08/24/2008            8   
1    Michael Phelps  19.0  2004            08/29/2004            6   
2    Michael Phelps  27.0  2012            08/12/2012            4   
3  Natalie Coughlin  25.0  2008            08/24/2008            1   
4  Natalie Coughlin  21.0  2004            08/29/2004            2   

   Silver Medals  Bronze Medals  Total Medals        Country  
0              0              0             8  United States  
1              0              2             8  United States  
2              2              0             6  United States  
3              2              3             6  United States  
4              2              1             5  United States

In [39]:
data_main_country.shape

(8657, 9)

In [49]:
data_main_country[data_main_country["Athlete"] == "Aleksandar Ciric"]

Athlete   Age  Year Closing Ceremony Date  Gold Medals  \
1491  Aleksandar Ciric  30.0  2008            08/24/2008            0   
1492  Aleksandar Ciric  26.0  2004            08/29/2004            0   
1493  Aleksandar Ciric  22.0  2000            10/01/2000            0   

      Silver Medals  Bronze Medals  Total Medals Country  
1491              0              1             1  Serbia  
1492              1              0             1  Serbia  
1493              0              1             1  Serbia

el fichero que relacion deberia tener una relacion de uno a uno, y no de una a muchos por que en este caso cuando se mezclan los dataframe se duplicara los datos.
Para resolver eso agregaremos `data_country_dp = data_country.drop_duplicates(subset='Athlete')`.

In [42]:
data_country_dp = data_country.drop_duplicates(subset = "Athlete")

In [45]:
print(f"El tamaño del data_country_dp es {len(data_country_dp)}, minetras que el data_country es {len(data_country)} \n")
print(f"existe una diferencia de {len(data_country)-len(data_country_dp)} entre el data_country y el data_country_dp")

El tamaño del data_country_dp es 6956, minetras que el data_country es 6970 

existe una diferencia de 14 entre el data_country y el data_country_dp


In [48]:
data_main_country = pd.merge(left = data_main, right = data_country_dp,
                            left_on = "Athlete", right_on = "Athlete")

len(data_main_country) == len(data_main)

True

In [53]:
data_sports_dp = data_sports.drop_duplicates(subset = "Athlete")

In [54]:
len(data_sports_dp) == len(a)

True

In [57]:
data_final = pd.merge(left = data_main_country, right= data_sports_dp,
                      left_on = "Athlete", right_on = "Athlete")

In [58]:
data_final.head()

Athlete   Age  Year Closing Ceremony Date  Gold Medals  \
0    Michael Phelps  23.0  2008            08/24/2008            8   
1    Michael Phelps  19.0  2004            08/29/2004            6   
2    Michael Phelps  27.0  2012            08/12/2012            4   
3  Natalie Coughlin  25.0  2008            08/24/2008            1   
4  Natalie Coughlin  21.0  2004            08/29/2004            2   

   Silver Medals  Bronze Medals  Total Medals        Country     Sport  
0              0              0             8  United States  Swimming  
1              0              2             8  United States  Swimming  
2              2              0             6  United States  Swimming  
3              2              3             6  United States  Swimming  
4              2              1             5  United States  Swimming

In [60]:
len(data_main) == len(data_final)

True

La cantidad de datos del data frame final es igual a la cantidad de datos del data frame inicial, sin tener el problema de que algunos atletas participan en distintos deportes, para otros paises.